py_hamt/sharded_zarr_store.py

```
@@ -429,39 +1044,43 @@ async def _load_or_initialize_shard_cache(
+                raise RuntimeError(
+                    f"Shard {shard_idx} not found in cache after pending load completed."
+                )
             except asyncio.TimeoutError:
-                # Clean up the pending load to allow retry
-                if shard_idx in self._pending_shard_loads:
-                    self._pending_shard_loads[shard_idx].set()
-                    del self._pending_shard_loads[shard_idx]
+                if cache_key in self._pending_shard_loads:
+                    self._pending_shard_loads[cache_key].set()
+                    del self._pending_shard_loads[cache_key]
                 raise RuntimeError(f"Timeout waiting for shard {shard_idx} to load.")
```

Ruff (0.15.20)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)